### ISSUE TITLE
custom gradient descent with momentum

### ISSUE THREAD

**Issue by nhejazi** [tlverse/hal9001#27]
Currently, we're implementing the LASSO by hand, including standard gradient descent. Theory has shown that faster convergence is possible using methods that incorporate momentum. Numerical experiments have shown that Nesterov momentum can be applied to the LASSO fairly easily, and with good results (see slides 27-29 of [this deck](https://www.princeton.edu/~yc5/ele538b_sparsity/lectures/lasso_algorithm_extension.pdf) for details). @jeremyrcoyle Would it be worthwhile for us to add in this method as an option and assess whether this would improve the performance of HAL?

**Comment by jeremyrcoyle** [comment]
We're currently doing coordinate descent, not gradient descent, so we don't have a step size/learning rate parameter. I'm not sure how Nesterov would translate to CD. My intuition is that CD will beat out GD even with momentum if the sparsity is high enough, which in our case it is. I suspect someone has already done this experiment for lasso in general, so maybe we can try to find that.

**Comment by nhejazi** [comment]
Ah, ok, I knew we had coordinate descent in there but assumed we also had gradient descent for some reason (I guess as an alternative). I don’t think the Nesterov idea could be applied in any useful way to coordinate descent given the formulation, and I believe your intuition that coordinate descent could be faster in some sparse regime. The slide deck I linked to gives a plot at the end comparing performance of vanilla gradient descent with Nesterov for sparse Lasso, but this of course doesn’t mean anything wrt coordinate descent. If we want to pursue this idea, I’m sure there’s plenty of information  we could draw from [_Statistical Learning with Sparsity_](https://web.stanford.edu/~hastie/StatLearnSparsity/), though proofs in these settings tend to be a bit assumption-heavy.

**Comment by nhejazi** [comment]
Stale -- closing since we've deprecated `lassi` as of v0.4.0+
